### PR TITLE
Document and test supported versions of Go

### PR DESCRIPTION
**Pull request by dmathieu**
Fixes https://github.com/open-telemetry/opentelemetry-proto-go/issues/195

### CODE REVIEW

**Comment by dmathieu** [comment]
I'm pondering adding a compatibility statement similar to the otel-go one.
https://github.com/open-telemetry/opentelemetry-go/tree/main?tab=readme-ov-file#compatibility

We don't really publish new releases with new Go versions though, and since this code is fully auto-generated, this may be a bit too much.

**Review by pellared — APPROVED** [review]
I think that we should have compatibility tests for all supported versions of Go (currently 1.22 and 1.23). These tests can be run on Ubuntu runner as there is no OS-specific code. 

Right now there is nothing that checks if the generated code actually builds.

File: internal/tools/go.mod [hunk 1]
```diff
@@ -1,5 +1,5 @@
 module go.opentelemetry.io/proto/internal/tools
 
-go 1.16
+go 1.22.0-
```
Comment (pellared) [inline internal/tools/go.mod:3]:
```suggestion
go 1.22.0
```

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        go-version: ["1.23.0", "~1.22.4"]
+        platform:
+          - os: ubuntu-latest
+            arch: "386"
+          - os: ubuntu-latest
+            arch: amd64
+          - os: windows-latest
+            arch: "386"
+          - os: windows-latest
+            arch: amd64
```
Comment (pellared) [inline .github/workflows/ci.yaml]: I think it is fine to just test on `ubuntu-latest`. No OS specific code here.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        with:
+          submodules: true
+      - name: Generate
+        run: make clean protobuf
+      - name: Check Clean Work Tree
+        run: make check-clean-work-tree
+  compatibility-test:
+    strategy:
+      matrix:
+        go-version: ["1.23.0", "~1.22.4"]
```
Comment (pellared) [inline .github/workflows/ci.yaml:28]: I think we should use the latest versions (as these are supported by the Go team)

```suggestion
        go-version: ["~1.23.0", "~1.22.4"]
```

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      uses: actions/checkout@v4
+    - name: Install Go
+      uses: actions/setup-go@v5
+      with:
+        go-version: ${{ matrix.go-version }}
+        check-latest: true
+        cache-dependency-path: "**/go.sum"
+    - name: Run tests
+      env:
+        GOARCH: ${{ matrix.platform.arch }}
```
Comment (pellared) [inline .github/workflows/ci.yaml:58]: IMO no need for that. No arch specific code

**Comment by dmathieu** [comment]
Once this is merged, I will make the `compatibility-test` check required.

**Review by pellared — APPROVED** [review]

**Comment by pellared** [comment]
> Once this is merged, I will make the `compatibility-test` check required.

I think you can do it even before merging 😉

**Comment by dmathieu** [comment]
It would prevent any other PR from being merged before this one.

**Review by pellared — CHANGES_REQUESTED** [review]
Something is wrong. `go.sum` files should not be empty.

**Comment by dmathieu** [comment]
Their size has vastly reduced. They are not empty though.
Their content is retrieved from running `go mod tidy`.

**Review by pellared — APPROVED** [review]

**Pull request merged**